### PR TITLE
Adds support for String(x::Symbol) = string(x)

### PR DESCRIPTION
**Pull request by dbeach24**
Bug fix for #16997 
defines String(x::Symbol) = string(x)

(This one's against master!)

### CODE REVIEW

**Comment by tkelman** [comment]
would have been less noise to rebase the existing PR https://github.com/JuliaLang/julia/pull/18120, but this is fine. will have to squash out that unnecessary merge commit

**Comment by musm** [comment]
string or print_to_string ?

File: base/strings/basic.jl [hunk 1]
```diff
@@ -34,6 +34,7 @@ If you need to subsequently modify `v`, use `String(copy(v))` instead.
 """
 String(v::Array{UInt8,1})
 
+String(x::Symbol) = string(x)
```
Comment (stevengj) [inline base/strings/basic.jl]: Should be more efficient to do `String(x::Symbol) = unsafe_string(Cstring(x))`, which copies directly from the underlying data.  (Seems to be about 3x faster in a quick benchmark.)

**Comment by stevengj** [comment]
Various places in the source code that use `string(somesymbol)` should also be updated to use `String(somesymbol)`, in the interest of eating our own dogfood.   A quick search finds at least the following cases:

* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/atomics.jl#L298
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/c.jl#L471
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/cartesian.jl#L274
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/cartesian.jl#L280
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/libdl.jl#L94
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/libdl.jl#L107
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/libdl.jl#L149
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/libuv.jl#L67
* all of the `string(mod)` calls in `loading.jl`
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/math.jl#L134
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/math.jl#L134
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/profile.jl#L642
* `string(name)` at https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/replutil.jl#L110
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/replutil.jl#L110
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/replutil.jl#L110
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/replutil.jl#L110
* https://github.com/JuliaLang/julia/blob/807047d07b1ece69884703fadf506368e3657d30/base/replutil.jl#L110

**Comment by stevengj** [comment]
Darn it, I screwed up my copy-and-paste in the above comment, so there are a number of duplicate lines that were supposed to be other occurrences of `string(somesymbol)`.  Oh well, we don't have to get every single spot.

File: base/strings/basic.jl (hunk 1 above)
Comment (TotalVerb) [inline base/strings/basic.jl]: Can we make this `convert(::Type{String}, x::Symbol)` for consistency? We already have `convert(Symbol, ::String)`. It would be weird if `Symbol(str)` and `s::Symbol = str` works, and `String(sym)` works, but `x::String = sym` does not.

**Review by TotalVerb — COMMENTED** [review]
I like this change. But this should probably be a method of `convert` (see comment).

File: base/strings/basic.jl (hunk 1 above)
Comment (StefanKarpinski) [inline base/strings/basic.jl]: Sure. Want to make a PR for that change?

Comment (TotalVerb) [inline base/strings/basic.jl]: Then we should merge this first, and make another PR to replace it with `convert`?

Comment (stevengj) [inline base/strings/basic.jl]: I'd prefer a single PR.  Ideally @dbeach24 would update this PR, but he doesn't seem to be around so maybe we should fork his branch and push a new PR?

Comment (dbeach24) [inline base/strings/basic.jl]: Been rather busy.  I'll work on it later this weekend.

Comment (stevengj) [inline base/strings/basic.jl]: Ah, thanks @dbeach24.

Comment (TotalVerb) [inline base/strings/basic.jl]: As far as I know, `Symbol`s are currently never freed, so this is safe. But I don't know if it's a good idea to depend on this behaviour. @stevengj

**Comment by dbeach24** [comment]
Just updated the PR with a new commit.
This implements `convert(::Type{String}, x::Symbol)`.

File: base/strings/basic.jl [hunk 2]
```diff
@@ -62,6 +62,7 @@ convert(::Type{Array{UInt8}}, s::AbstractString) = String(s).data
 convert(::Type{String}, s::AbstractString) = String(s)
 convert(::Type{Vector{Char}}, s::AbstractString) = collect(s)
 convert(::Type{Symbol}, s::AbstractString) = Symbol(s)
+convert(::Type{String}, s::Symbol) = unsafe_wrap(String, Cstring(s))
```
Comment (dbeach24) [inline base/strings/basic.jl]: This is a very efficient conversion, but I need a ruling on the safety of doing this w.r.t. memory management of symbols.

Comment (tkelman) [inline base/strings/basic.jl]: would this mean the `.data` field of the result of converting the same symbol to multiple different string objects would all point to the same data?

Comment (stevengj) [inline base/strings/basic.jl]: @tkelman, no, because `unsafe_wrap` makes a copy.

Comment (TotalVerb) [inline base/strings/basic.jl]: They would point to the same data, but the `.data` array would be different. They would be distinct `.data` arrays that share the same underlying data.

Comment (stevengj) [inline base/strings/basic.jl]: Sorry, yes, this would means that all the strings have data that points to the same memory.

My suggestion is to use `unsafe_string` here, not, `unsafe_wrap`.   I don't think the extra efficiency of not making a copy is worth it for the danger of being able to modify `Symbol` contents through the `data` field of a string.

Comment (dbeach24) [inline base/strings/basic.jl]: @stevengj I will follow your advice.

It seems that your view includes two kinds of immutability.  e.g. Symbols are IMMUTABLE, while Strings are merely immutable... and we can't go trusting an immutable object to hold the definitive copy of IMMUTABLE data.

I am correctly understanding this?

Comment (TotalVerb) [inline base/strings/basic.jl]: It's not like someone can't muck around with the `Symbol` contents anyway using `Cstring` directly, though. Mutating the `data` field of a string should be a no-go unless you make the string yourself.

Comment (dbeach24) [inline base/strings/basic.jl]: If we all agree that mutating "someone else's" String is a no-go, then I'd be more inclined to stick with the `unsafe_wrap` version.

Comment (stevengj) [inline base/strings/basic.jl]: It's true that you're not supposed to mutate strings, and `unsafe_wrap` should be safe for symbol data if people are following the rules, but I just wanted to reduce the temptation here.  I'm fine either way, however.

Comment (dbeach24) [inline base/strings/basic.jl]: There's probably not enough of a performance difference to matter.  I'll go with the "safer" unsafe version.

```jl
julia> symbols = [gensym() for i=1:10^6];
julia> f1(symbols) = for s in symbols unsafe_wrap(String, Cstring(s)) end
julia> f2(symbols) = for s in symbols unsafe_string(Cstring(s)) end

julia> @time f1(symbols)
  0.040287 seconds (2.00 M allocations: 91.553 MB, 14.31% gc time)

julia> @time f2(symbols)
  0.052429 seconds (2.00 M allocations: 106.812 MB, 15.83% gc time)
```

**Comment by KristofferC** [comment]
Restarted travis. Previous error was the broadcast issue. LGTM when CI pass.

**Review by KristofferC — APPROVED** [review]

**Comment by tkelman** [comment]
What is
```
From worker 7:	     * broadcast            Error During Test
	From worker 7:	  Test threw an exception of type ErrorException
	From worker 7:	  Expression: broadcast(+,1.0,(0,-2.0)) == (1.0,-1.0)
	From worker 7:	  type #+ has no field contents
	From worker 7:	   in #35 at ./broadcast.jl:260 [inlined]
	From worker 7:	   in _ntuple at ./tuple.jl:101 [inlined]
	From worker 7:	   in ntuple(::Base.Broadcast.##35#36, ::Type{Val{2}}) at ./tuple.jl:94
	From worker 7:	   in macro expansion at ./broadcast.jl:260 [inlined]
	From worker 7:	   in broadcast_tup(::Base.#+, ::Tuple{Float64,Tuple{Int64,Float64}}, ::Type{Val{2}}, ::Int64) at ./broadcast.jl:259
	From worker 7:	   in broadcast_c(::Function, ::Type{Tuple}, ::Float64, ::Vararg{Any,N}) at ./broadcast.jl:269
	From worker 7:	   in broadcast(::Function, ::Float64, ::Tuple{Int64,Float64}) at ./broadcast.jl:340
	From worker 7:	   in include_string(::String, ::String) at ./loading.jl:478
	From worker 7:	   in include_from_node1(::String) at ./loading.jl:544
	From worker 7:	   in macro expansion at ./util.jl:232 [inlined]
	From worker 7:	   in runtests(::String) at /tmp/julia/share/julia/test/testdefs.jl:7
	From worker 7:	   in (::Base.##619#621{Base.CallMsg{:call_fetch}})() at ./multi.jl:1425
	From worker 7:	   in run_work_thunk(::Base.##619#621{Base.CallMsg{:call_fetch}}, ::Bool) at ./multi.jl:1005
	From worker 7:	   in macro expansion at ./multi.jl:1425 [inlined]
	From worker 7:	   in (::Base.##618#620{Base.CallMsg{:call_fetch},Base.MsgHeader,TCPSocket})() at ./event.jl:68
```
anyone seen that before?

**Comment by TotalVerb** [comment]
@tkelman It's https://github.com/JuliaLang/julia/issues/18577.

**Comment by tkelman** [comment]
Oh right, then I guess Kristoffer didn't restart all jobs.

**Comment by KristofferC** [comment]
Looks like I only restarded one of them. Oops.

**Comment by kshyatt** [comment]
@KristofferC OKed this and CI passed so I'm going to merge.
